classes/04class/iac/README.md

```diff
@@ -0,0 +1,343 @@
+# Infrastructure as Code (IaC)
+This class is an introduction to Infrastructure as Code (IaC).
+The main goal is to explain the purpose, the benefits and how to use it.
+In this class, we'll be using Terraform as the main tool to provision resources in AWS through code.
+
+***Contents***
+- [What is IaC?](#what-is-iac)
+  - [Benefits of IaC](#benefits-of-iac)
+  - [IaC Models](#iac-models)
+  - [IaC Methods](#iac-methods)
+- [Terraform Foundation](#terraform-foundation)
+  - [Terraform Providers](#terrform-providers)
+  - [Terraform State](#terraform-state)
+- [Terraform Client](#terraform-client)
+ - [Terraform Resources](#terraform-resources)
+ - [The Power of Input Variables](#the-power-of-input-variables)
+ - [Outputs](#outputs)
+ - [File Organization](#file-organization)
+ - [Running some code](#running-some-code)
+ - [Extra](#extra)
+  - [Modules](#modules)
+ - [Appendix](#appendix)
+
+
+## What is IaC?
+Infrastructure as Code (IaC) is the process of create, configure and manage IT resources through machine-readable file. Instead of manually creating and configuring IT resources, it's possible to have everything created in an automated and repeatable way.
+
+IaC won't replace physical installation (racking and cabling equipments, for example), but it can be used for most of the task executed after that.
+
+As an example, all the configuration that was executed to create the VPC, subnets, routing tables during exercises from class 2 or even configuring ECS during exercises in class 3 could be coded.
+
+With all those manual steps in a code format, it would be possible to create those resources multiple in a matter of minutes, instead of going into multiple clicks in the AWS console.
+
+### Benefits of IaC
+The benefits of IaC are normally around three pillars:
+
+#### Cost
+Repeatable tasks that are transformed into code can reduce costs. Instead of spending 2 hours creating an environment manually that needs to be created weekly, though code, that environment creation can take less than 5 minutes. 
+
+With that, you're not only reducing the cost to deliver that environment, but the engineers will be able to focus on more critical things for the company.
+
+#### Speed
+IaC will make the resources available faster, enabling developers to start working sooner. Especially for environments with several changes or with fast innovation pace, where new environments need to be created frequently, IaC can enable this fast pace.
+
+#### Risk
+Because everything is defined through code, you can ensure that no manual mistake can be made during any of the configuration. Also, you'll know what is running on your environment, considering that deployments are only made through code, and no manual changes are possible.
```

Comment:
> no manual changes are possible.

manual changes are possible, but much easier to restore the system to a known infra state